bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoHandler.java

```diff
@@ -296,135 +324,154 @@ public void onValueReceived(@Nullable String variable, @Nullable String value, @
+            UpnpIOService localservice = service;
+            if (localservice != null) {
+                if (localservice.isRegistered(this)) {
+                    logger.debug("Removing WeMo GENA subscription for '{}'", this);
+                    ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+                    String subscription = BASICEVENT + "1";
+
+                    if (subscriptionState.get(subscription) != null) {
+                        logger.debug("WeMo {}: Unsubscribing from service {}...", getUDN(), subscription);
+                        localservice.removeSubscription(this, subscription);
+                    }
 
-            if (thingTypeUID.equals(THING_TYPE_INSIGHT)) {
-                subscription = "insight1";
-                if (subscriptionState.get(subscription) != null) {
-                    logger.debug("WeMo {}: Unsubscribing from service {}...", getUDN(), subscription);
-                    service.removeSubscription(this, subscription);
+                    if (THING_TYPE_INSIGHT.equals(thingTypeUID)) {
+                        subscription = INSIGHTEVENT + "1";
+                        if (subscriptionState.get(subscription) != null) {
+                            logger.debug("WeMo {}: Unsubscribing from service {}...", getUDN(), subscription);
+                            localservice.removeSubscription(this, subscription);
+                        }
+                    }
+                    subscriptionState = new HashMap<>();
+                    localservice.unregisterParticipant(this);
                 }
             }
-            subscriptionState = new HashMap<>();
-            service.unregisterParticipant(this);
-        }
-    }
-
-    private synchronized void onUpdate() {
-        ScheduledFuture<?> job = refreshJob;
-        if (job == null || job.isCancelled()) {
-            Configuration config = getThing().getConfiguration();
-            int refreshInterval = DEFAULT_REFRESH_INTERVALL_SECONDS;
-            Object refreshConfig = config.get("refresh");
-            if (refreshConfig != null) {
-                refreshInterval = ((BigDecimal) refreshConfig).intValue();
-            }
-            refreshJob = scheduler.scheduleWithFixedDelay(refreshRunnable, 0, refreshInterval, TimeUnit.SECONDS);
         }
     }
 
     private boolean isUpnpDeviceRegistered() {
-        return service.isRegistered(this);
+        UpnpIOService localservice = service;
+        if (localservice != null) {
+            return localservice.isRegistered(this);
+        }
+        return false;
     }
 
     @Override
     public String getUDN() {
         return (String) this.getThing().getConfiguration().get(UDN);
     }
 
+    public String getHost() {
+        UpnpIOService localservice = service;
+        if (localservice != null) {
+            URL descriptorURL = localservice.getDescriptorURL(this);
+            if (descriptorURL != null) {
+                host = descriptorURL.getHost();
```

Comment:
return this value directly instead of setting the instance variable first. Method returns String, so caller should set instance variable, i.e.: `host = getHost();`. Currently return value is not used, and it's counter-intuitive that a get method actually sets something. Seems also that it should be private?

Reply:
changet, but cannot be private, otherwise we cannot assert a value in Test Classes....

Those changes seem to have eliminated the failing OSGITests, still WemoHanlerTest is failing, but we are coming closer....

Reply:
Hmm, I'd rather treat that as a clue than a fix. Asserting on host should not fix the tests as it will just check one more thing. But it will cause that method to be called though, so if this changes the result, it indicates that the test didn't reach the code where host would be set in normal way. Not sure if this helps, but seems you are on the right track.